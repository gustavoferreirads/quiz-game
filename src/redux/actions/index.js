export const QUESTIONS = 'QUESTIONS';
export const ANSWER = 'ANSWER';
