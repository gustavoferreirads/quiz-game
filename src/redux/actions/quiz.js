import axios from 'axios';
import { createAction } from 'redux-actions';
import { ANSWER, QUESTIONS } from './index';

export function doAnswer(answer) {
  return dispatch => {
    dispatch({
      type: ANSWER,
      answer,
    });
  };
}

function get(amount = 10, difficulty = 'hard', type = 'boolean') {
  return axios.get('api.php', {
    params: {
      amount,
      difficulty,
      type,
    },
  });
}
export const fetchQuestions = createAction(QUESTIONS, get);
