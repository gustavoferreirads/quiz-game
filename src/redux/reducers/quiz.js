import { AllHtmlEntities } from 'html-entities';
import { ANSWER, QUESTIONS } from '../actions';

const initialState = {
  questions: [],
  size: 0,
  current: { index: -1, wrapQuestion: {} },
  answers: [],
  score: 0,
  percent: 0,
};

const generateResult = state => {
  const { questions } = state;
  let score = 0;
  const answers = questions.map((question, index) => {
    const isCorrect = state.answers[index] === question.correct_answer;
    score = isCorrect ? score + 1 : score;
    return { ...question, isCorrect };
  });
  const percent = (score / answers.length) * 100;
  return { ...initialState, score, answers, percent };
};

const nextQuestion = state => {
  const { current, questions } = state;

  const index = current.index + 1;

  if (index < questions.length) {
    const wrapQuestion = questions[index];

    return {
      ...state,
      current: { index, wrapQuestion, position: index + 1 },
    };
  }

  return generateResult(state);
};

const reducer = (state = initialState, action) => {
  if (action.type === QUESTIONS) {
    const { payload } = action;
    const questions = payload.data.results.map(q => ({
      ...q,
      question: new AllHtmlEntities().decode(q.question),
    }));
    const newState = { ...initialState, questions, size: questions.length };
    return nextQuestion(newState);
  }

  if (action.type === ANSWER) {
    const { answers } = state;

    answers.push(action.answer);

    return nextQuestion({
      ...state,
      answers,
    });
  }

  return state;
};

export default reducer;
