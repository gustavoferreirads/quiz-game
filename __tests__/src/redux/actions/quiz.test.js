import mockAxios from 'axios';
import mockStore from 'redux-mock-store';
import { fetchQuestions, doAnswer } from '../../../../src/redux/actions/quiz';

const store = mockStore();

beforeEach(() => {
  store.clearActions();
});

it('fetch questions', async () => {
  mockAxios.get.mockImplementationOnce(() =>
    Promise.resolve({
      data: { results: [{ id: 1, question: 'question' }] },
    }),
  );
  await store.dispatch(fetchQuestions());
  expect(store.getActions()).toMatchSnapshot();
});

it('do answer', () => {
  mockAxios.get.mockImplementationOnce(() =>
    Promise.resolve({
      data: { results: [{ id: 1, question: 'question' }] },
    }),
  );
  store.dispatch(doAnswer('True'));
  expect(store.getActions()).toMatchSnapshot();
});
