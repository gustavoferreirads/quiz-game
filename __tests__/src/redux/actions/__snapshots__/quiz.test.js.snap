// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`do answer 1`] = `
Array [
  Object {
    "answer": "True",
    "type": "ANSWER",
  },
]
`;

exports[`fetch questions 1`] = `
Array [
  Object {
    "payload": Promise {
      "_40": 0,
      "_55": Object {
        "data": Object {
          "results": Array [
            Object {
              "id": 1,
              "question": "question",
            },
          ],
        },
      },
      "_65": 1,
      "_72": null,
    },
    "type": "QUESTIONS",
  },
]
`;
